profiler/profile.go

```diff
@@ -236,22 +240,13 @@ func (l *limitedTraceCollector) Write(p []byte) (n int, err error) {
 		return
 	}
 	l.written += n
-	if l.written >= l.limit {
-		// We can't stop tracing within this method, because trace.Stop
-		// won't return until all pending events are written, which
-		// requires calling this function, meaning we'll deadlock.
-		go l.Stop()
+	if l.written >= l.limit && !l.stopped {
+		l.stopped = true
+		close(l.done)
```

Comment:
Q: This Write method assumes its not called concurrently, right?

Reply:
That's right. I'm actually not sure if an io.Writer could ever be safely used concurrently, come to think of it.